### ISSUE TITLE
NullReferenceException in StartReceive method

### ISSUE THREAD

**Issue by mikhail-khalizev** [tmds/Tmds.MDns#36]
Hi.

I recently started using Tmds.MDns `0.7.0` and often get a NullReferenceException:

```
System.NullReferenceException
  HResult=0x80004003
  Сообщение = Object reference not set to an instance of an object.
  Источник = Tmds.MDns
  Трассировка стека:
   at Tmds.MDns.NetworkInterfaceHandler.StartReceive()
   at Tmds.MDns.NetworkInterfaceHandler.OnReceive(Object sender, SocketAsyncEventArgs args)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped)

  Изначально это исключение было создано в этом стеке вызовов: 
    [Внешний код]
```

And also in `Send` method: [2020-05-10_21-58-07.png](https://yadi.sk/i/7-p-0i7K182DkQ)

Can you look and fix it?

**Comment by tmds** [comment]
Can you reproduce this and find out what is `null`?

**Comment by mikhail-khalizev** [comment]
No, it's been too long ...

**Comment by ossaltu** [comment]
Hi,
I am getting similar exception when calling StopBrowse():

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Tmds.MDns
  StackTrace:
   at Tmds.MDns.NetworkInterfaceHandler.StartReceive()
   at Tmds.MDns.NetworkInterfaceHandler.OnReceive(Object sender, SocketAsyncEventArgs args)
   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted(SocketAsyncEventArgs e)
   at System.Net.Sockets.SocketAsyncEventArgs.ExecutionCallback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncSuccess(Int32 bytesTransferred, SocketFlags flags)
   at System.Net.Sockets.SocketAsyncEventArgs.<>c.<.cctor>b__177_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)

  This exception was originally thrown at this call stack:
    [External Code]
```

There was no hint, on what is null, so I compiled the lib in debug and reproduced error again, received this exception:

```
System.NullReferenceException: 'Object reference not set to an instance of an object.'

_socket was null.

System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Tmds.MDns
  StackTrace:
   at Tmds.MDns.NetworkInterfaceHandler.StartReceive() in C:\Users\ossal\Downloads\Tmds.MDns-0.7.0\Tmds.MDns-0.7.0\src\Tmds.MDns\NetworkInterfaceHandler.cs:line 178
```

This exception shows up randomly in 1 of 10 runs. Lib version 7.0, target netstandard 1.3, lib is used in UWP application.
Let me know, if you need more info.

**Comment by tmds** [comment]
> target netstandard 1.3

This target is missing a `if (_socket == null)` check here:

https://github.com/tmds/Tmds.MDns/blob/01f0ba78c89da1f4e1929a2e8ba16f2d4a01fa33/src/Tmds.MDns/NetworkInterfaceHandler.cs#L196-L216

**Comment by ossaltu** [comment]
change 197 to

`if (args.SocketError != SocketError.Success || _socket == null)`

now getting exception in different line:

```
System.NullReferenceException: 'Object reference not set to an instance of an object.'

_socket was null.

System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Tmds.MDns
  StackTrace:
   at Tmds.MDns.NetworkInterfaceHandler.Send(IList`1 packets) in C:\Users\ossal\Downloads\Tmds.MDns-0.7.0\Tmds.MDns-0.7.0\src\Tmds.MDns\NetworkInterfaceHandler.cs:line 131

```

**Comment by tmds** [comment]
> now getting exception in different line:

@ossaltu I guess that's because the debugger is attached? It should get swallowed by the `catch` block:

https://github.com/tmds/Tmds.MDns/blob/01f0ba78c89da1f4e1929a2e8ba16f2d4a01fa33/src/Tmds.MDns/NetworkInterfaceHandler.cs#L127-L135

**Comment by ossaltu** [comment]
Yes, that's true. 
When can new version with that fix be expected?

PS. throw-catching around null ref exceptions is not good practice, when this can be solved with one "if". Mind the exception handling overhead.

**Comment by tmds** [comment]
Fixed in v0.7.1 which is now on nuget.org.
